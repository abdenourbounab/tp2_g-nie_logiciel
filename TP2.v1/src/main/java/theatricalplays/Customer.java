package theatricalplays;

public class Customer {
    private String name;
    private int customerNumber; 
    private int loyaltyPoints;

    public Customer(String name, int customerNumber, int loyaltyPoints) { 
        this.name = name;
        this.customerNumber = customerNumber;
        this.loyaltyPoints = loyaltyPoints;
    }

    // Getter for name
    public String getName() {
        return name;
    }

    // Setter for name
    public void setName(String name) {
        this.name = name;
    }

    // Getter for customerNumber
    public int getCustomerNumber() { 
        return customerNumber;
    }

    // Setter for customerNumber
    public void setCustomerNumber(int customerNumber) {
        this.customerNumber = customerNumber;
    }

    // Getter for loyaltyPoints
    public int getLoyaltyPoints() {
        return loyaltyPoints;
    }

    // Setter for loyaltyPoints
    public void setLoyaltyPoints(int loyaltyPoints) {
        this.loyaltyPoints = loyaltyPoints;
    }
}
