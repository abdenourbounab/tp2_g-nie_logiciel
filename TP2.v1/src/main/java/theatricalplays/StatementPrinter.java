package theatricalplays;

import java.text.NumberFormat;
import java.util.Locale;
import java.util.Map;


public class StatementPrinter {

  public String print(Invoice invoice, Map<String, Play> plays) {
    //int totalAmount = 0;
    double totalAmount = 0.0;
    int volumeCredits = 0;
    StringBuilder result = new StringBuilder();
    
    // En-tête de la déclaration
    result.append(String.format("Statement for %s\n", invoice.customer));

    NumberFormat frmt = NumberFormat.getCurrencyInstance(Locale.US);

    for (Performance perf : invoice.performances) {
      Play play = plays.get(perf.playID);
      //int thisAmount = 0;
      double thisAmount = 0.0;

      // Calcul du montant en fonction du type de pièce de théâtre
      switch (play.type) {

        case Play.PlayType.tragedy:
          thisAmount = 400.0;
          if (perf.audience > 30) {
            thisAmount += 10.0 * (perf.audience - 30);
          }
          break;

        case Play.PlayType.comedy:
          thisAmount = 300.0;
          if (perf.audience > 20) {
            thisAmount += 100.0 + 5.0 * (perf.audience - 20);
          }
          thisAmount += 3.0 * perf.audience;
          break;

        default:
          throw new Error("unknown type: ${play.type}");
      }

      // Ajouter des crédits de volume
      volumeCredits += Math.max(perf.audience - 30, 0);

      // Ajouter des crédits supplémentaires pour chaque groupe de 5 personnes pour les comédies
      if (Play.PlayType.comedy.equals(play.type)) volumeCredits += Math.floor(perf.audience / 5);

      // Ajouter une ligne pour cette représentation
      result.append(String.format("  %s: %s (%s seats)\n", play.name, frmt.format(thisAmount), perf.audience));
      totalAmount += thisAmount;
    }

    // Ajouter le montant total dû
    result.append(String.format("Amount owed is %s\n", frmt.format(totalAmount)));

     // Ajouter le total des crédits gagnés
    result.append(String.format("You earned %s credits\n", volumeCredits));

    return result.toString();
  }

}
