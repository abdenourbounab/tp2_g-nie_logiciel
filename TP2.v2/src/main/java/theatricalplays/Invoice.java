package theatricalplays;

import java.util.List;

public class Invoice {

  private Customer customer;
  public List<Performance> performances;

  public Invoice(Customer customer, List<Performance> performances) {
    this.customer = customer;
    this.performances = performances;
  }

  public Customer getCustomer() {
    return customer;
  }
  
}
