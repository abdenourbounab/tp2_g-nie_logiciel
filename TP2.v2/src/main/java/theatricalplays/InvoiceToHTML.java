package theatricalplays;

import java.text.NumberFormat;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class InvoiceToHTML {

    public static String generateHTMLInvoice(String customer, List<Performance> performances, Map<String, Play> plays) {
        // Entête de la page HTML
        StringBuilder html = new StringBuilder();
        html.append("<html> <style> table, th, td {border:1px solid black;} </style> <head><title>Facture</title></head><body>");

        // En-tête de la facture
        html.append("<h1>Invoice</h1>");

        html.append("<ul><li><b>Client : </b>").append(customer).append("</li></ul>");

        html.append("<table><tr><th>Piece</th><th>Seats sold</th><th>Price</th></tr>");

        // Initialize volumeCredits to 0
        int volumeCredits = 0;

        // Boucle sur les performances
        for (Performance perf : performances) {
            Play play = plays.get(perf.playID);

            // Calcul du montant et des crédits pour cette performance
            double thisAmount = calculatePerformanceAmount(play, perf);
            int thisVolumeCredits = calculateVolumeCredits(play, perf);

            // Ajouter des crédits pour cette performance
            volumeCredits += thisVolumeCredits;

            // Ajouter une ligne à la table HTML
            html.append("<tr>")
                .append("<td>").append(play.name).append("</td>")
                .append("<td>").append(perf.audience).append("</td>")
                .append("<td>").append(NumberFormat.getCurrencyInstance(Locale.US).format(thisAmount)).append("</td>")
                .append("</tr>");
        }

        

        // Calcul du montant total
        double totalAmount = calculateTotalAmount(performances, plays);

        // Ajout du montant total
        html.append("<tr><th colspan=\"2\">Total owed : </th>").append("<td>").append(NumberFormat.getCurrencyInstance(Locale.US).format(totalAmount)).append("</td></tr>");

        // Ajout des volumeCredits
        html.append("<tr><th colspan=\"2\">Fidelity points earned : </th>").append("<td>").append(volumeCredits).append("</td></tr>");

        
        
        html.append("</table>");
        // Fin de la page HTML
        html.append("</body></html>");

        return html.toString();
    }

    private static double calculatePerformanceAmount(Play play, Performance perf) {
        double thisAmount = 0.0;
        switch (play.type) {
            case Play.PlayType.tragedy:
                thisAmount = 400.0;
                if (perf.audience > 30) {
                    thisAmount += 10.0 * (perf.audience - 30);
                }
                break;
            case Play.PlayType.comedy:
                thisAmount = 300.0;
                if (perf.audience > 20) {
                    thisAmount += 100.0 + 5.0 * (perf.audience - 20);
                }
                thisAmount += 3.0 * perf.audience;
                break;
            default:
                throw new Error("unknown type: " + play.type);
        }
        return thisAmount;
    }

    private static int calculateVolumeCredits(Play play, Performance perf) {
        // Ajouter des crédits de volume
        int volumeCredits = Math.max(perf.audience - 30, 0);

        // Ajouter des crédits supplémentaires pour chaque groupe de 5 personnes pour les comédies
        if (play.type.equals(Play.PlayType.comedy)) {
            volumeCredits += perf.audience / 5;
        }
        return volumeCredits;
    }

    private static double calculateTotalAmount(List<Performance> performances, Map<String, Play> plays) {
        double totalAmount = 0.0;
        for (Performance perf : performances) {
            Play play = plays.get(perf.playID);
            totalAmount += calculatePerformanceAmount(play, perf);
        }
        return totalAmount;
    }

    
}

