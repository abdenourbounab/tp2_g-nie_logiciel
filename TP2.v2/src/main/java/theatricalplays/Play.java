package theatricalplays;

public class Play {
  public enum PlayType {
    tragedy,
    comedy,
    history,
    pastoral
  }

  public String name;
  public PlayType type;

  // Private constructor
  private Play(String name, PlayType type) {
    this.name = name;
    this.type = type;
  }

  // Static factory method to create Play objects
  public static Play createPlay(String name, PlayType type) {
    return new Play(name, type);
  }
}
