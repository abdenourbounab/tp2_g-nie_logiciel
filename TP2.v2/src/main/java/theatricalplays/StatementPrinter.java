package theatricalplays;

import java.text.NumberFormat;
import java.util.Locale;
import java.util.Map;


public class StatementPrinter {

  public String print(Invoice invoice, Map<String, Play> plays) {
    //int totalAmount = 0;
    double totalAmount = 0.0;
    int volumeCredits = 0;
    StringBuilder result = new StringBuilder();
    
    // En-tête de la déclaration
    result.append(String.format("Statement for %s\n", invoice.getCustomer().getName()));

    NumberFormat frmt = NumberFormat.getCurrencyInstance(Locale.US);

    for (Performance perf : invoice.performances) {
      Play play = plays.get(perf.playID);
      //int thisAmount = 0;
      double thisAmount = 0.0;

      // Calcul du montant en fonction du type de pièce de théâtre
      switch (play.type) {

        case Play.PlayType.tragedy:
          thisAmount = 400.0;
          if (perf.audience > 30) {
            thisAmount += 10.0 * (perf.audience - 30);
          }
          break;

        case Play.PlayType.comedy:
          thisAmount = 300.0;
          if (perf.audience > 20) {
            thisAmount += 100.0 + 5.0 * (perf.audience - 20);
          }
          thisAmount += 3.0 * perf.audience;
          break;

        default:
          throw new Error("unknown type: ${play.type}");
      }

      // Ajouter des crédits de volume
      volumeCredits += Math.max(perf.audience - 30, 0);

      // Ajouter des crédits supplémentaires pour chaque groupe de 5 personnes pour les comédies
      if (Play.PlayType.comedy.equals(play.type)) volumeCredits += Math.floor(perf.audience / 5);

      // Ajouter une ligne pour cette représentation
      result.append(String.format("  %s: %s (%s seats)\n", play.name, frmt.format(thisAmount), perf.audience));
      totalAmount += thisAmount;
    }
    
    invoice.getCustomer().setLoyaltyPoints(invoice.getCustomer().getLoyaltyPoints() + volumeCredits);

    // Ajouter le montant total dû
    result.append(String.format("Amount owed is %s\n", frmt.format(totalAmount)));

     // Ajouter le total des crédits gagnés
    result.append(String.format("You earned %s credits\n", volumeCredits));

    
    int customerLoyaltyPoints = invoice.getCustomer().getLoyaltyPoints();
    if (customerLoyaltyPoints >= 150) {
      double totalAmountReduced = totalAmount - 15.0;

      result.append(String.format("You have %s Loyalty Points, you get a $15 reduction\n", customerLoyaltyPoints));

      customerLoyaltyPoints -= 150;
      invoice.getCustomer().setLoyaltyPoints(customerLoyaltyPoints);

      result.append(String.format("Amount owed after reduction %s\n", frmt.format(totalAmountReduced)));
      result.append(String.format("New Fidelity Points %s\n", customerLoyaltyPoints));

    }

    return result.toString();

}

}
