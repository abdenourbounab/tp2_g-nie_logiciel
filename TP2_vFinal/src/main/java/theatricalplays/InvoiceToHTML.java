package theatricalplays;

import java.text.NumberFormat;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class InvoiceToHTML {

    public static String generateHTMLInvoice(Customer customer, List<Performance> performances, Map<String, Play> plays) {
        // Entête de la page HTML
        StringBuilder html = new StringBuilder();
        html.append("<html> <style> table, th, td {border:1px solid black;} </style> <head><title>Facture</title></head><body>");

        // En-tête de la facture
        html.append("<h1>Invoice</h1>");

        html.append("<ul><li><b>Client : </b>").append(customer.getName()).append("</li></ul>");

        html.append("<table><tr><th>Piece</th><th>Seats sold</th><th>Price</th></tr>");

        // Initialize volumeCredits to 0
        int volumeCredits = 0;

        // Boucle sur les performances
        for (Performance perf : performances) {
            Play play = plays.get(perf.playID);

            // Calcul du montant et des crédits pour cette performance
            double thisAmount = calculatePerformanceAmount(play, perf);
            int thisVolumeCredits = calculateVolumeCredits(play, perf);

            // Ajouter des crédits pour cette performance
            volumeCredits += thisVolumeCredits;

            // Ajouter une ligne à la table HTML
            html.append("<tr>")
                .append("<td>").append(play.name).append("</td>")
                .append("<td>").append(perf.audience).append("</td>")
                .append("<td>").append(NumberFormat.getCurrencyInstance(Locale.US).format(thisAmount)).append("</td>")
                .append("</tr>");
        }

        

        // Calcul du montant total
        double totalAmount = calculateTotalAmount(performances, plays);

        customer.setLoyaltyPoints(customer.getLoyaltyPoints() + volumeCredits);

        // Ajout du montant total
        html.append("<tr><th colspan=\"2\">Total owed : </th>").append("<td>").append(NumberFormat.getCurrencyInstance(Locale.US).format(totalAmount)).append("</td></tr>");

        // Ajout des volumeCredits
        html.append("<tr><th colspan=\"2\">Fidelity points earned : </th>").append("<td>").append(volumeCredits).append("</td></tr>");

        int customerLoyaltyPoints = customer.getLoyaltyPoints();

        if (customerLoyaltyPoints >= 150) {
            
            double discountAmount = 15.0;
            double totalAmountReduced = totalAmount - discountAmount;
            // Ajout du message de réduction
            html.append("<tr><th colspan=\"3\">You have ").append(customerLoyaltyPoints)
                .append(" Loyalty Points, you get a $15 reduction</th></tr>");

            customerLoyaltyPoints -= 150;
            customer.setLoyaltyPoints(customerLoyaltyPoints);

            // Ajout du montant après réduction
            html.append("<tr><th colspan=\"2\">Amount owed after reduction : </th>")
                .append("<td>").append(NumberFormat.getCurrencyInstance(Locale.US).format(totalAmountReduced)).append("</td></tr>");

            // Ajout des nouveaux points de fidélité
            html.append("<tr><th colspan=\"2\">New Fidelity Points : </th>")
                .append("<td>").append(customerLoyaltyPoints).append("</td></tr>");
        }


        
        html.append("</table>");
        // Fin de la page HTML
        html.append("</body></html>");

        return html.toString();
    }

    private static double calculatePerformanceAmount(Play play, Performance perf) {
        double thisAmount = 0.0;
        switch (play.type) {
            case Play.PlayType.tragedy:
                thisAmount = 400.0;
                if (perf.audience > 30) {
                    thisAmount += 10.0 * (perf.audience - 30);
                }
                break;
            case Play.PlayType.comedy:
                thisAmount = 300.0;
                if (perf.audience > 20) {
                    thisAmount += 100.0 + 5.0 * (perf.audience - 20);
                }
                thisAmount += 3.0 * perf.audience;
                break;
            default:
                throw new Error("unknown type: " + play.type);
        }
        return thisAmount;
    }

    private static int calculateVolumeCredits(Play play, Performance perf) {
        // Ajouter des crédits de volume
        int volumeCredits = Math.max(perf.audience - 30, 0);

        // Ajouter des crédits supplémentaires pour chaque groupe de 5 personnes pour les comédies
        if (play.type.equals(Play.PlayType.comedy)) {
            volumeCredits += perf.audience / 5;
        }
        return volumeCredits;
    }

    private static double calculateTotalAmount(List<Performance> performances, Map<String, Play> plays) {
        double totalAmount = 0.0;
        for (Performance perf : performances) {
            Play play = plays.get(perf.playID);
            totalAmount += calculatePerformanceAmount(play, perf);
        }
        return totalAmount;
    }

    
}

