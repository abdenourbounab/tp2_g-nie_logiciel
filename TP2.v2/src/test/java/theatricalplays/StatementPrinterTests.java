package theatricalplays;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertEquals;
import theatricalplays.InvoiceToHTML;

import java.util.List;
import java.util.Map;

import static org.approvaltests.Approvals.verify;


public class StatementPrinterTests {

    @Test
    void exampleStatement() {
        Map<String, Play> plays = Map.of(
                "hamlet",  Play.createPlay("Hamlet", Play.PlayType.tragedy),
                "as-like", Play.createPlay("As You Like It", Play.PlayType.comedy),
                "othello", Play.createPlay("Othello", Play.PlayType.tragedy));

        Customer customer = new Customer("BigCo", 12345, 100); // Create a Customer object

        Invoice invoice = new Invoice(customer, List.of(
                new Performance("hamlet", 55),
                new Performance("as-like", 35),
                new Performance("othello", 40)));

        StatementPrinter statementPrinter = new StatementPrinter();
        var result = statementPrinter.print(invoice, plays);

        verify(result);
    }

    @Test
    void generateHTMLInvoiceTest() {
        Map<String, Play> plays = Map.of(
                "hamlet", Play.createPlay("Hamlet", Play.PlayType.tragedy),
                "as-like", Play.createPlay("As You Like It", Play.PlayType.comedy),
                "othello", Play.createPlay("Othello", Play.PlayType.tragedy));

        Customer customer = new Customer("BigCo", 12345, 100); // Create a Customer object

        Invoice invoice = new Invoice(customer, List.of(
                new Performance("hamlet", 55),
                new Performance("as-like", 35),
                new Performance("othello", 40)));

        var generatedHTML =  InvoiceToHTML.generateHTMLInvoice(invoice.getCustomer().getName(), invoice.performances, plays); // Use the updated method with Invoice object

        verify(generatedHTML);
    }

    @Test
    void tragedyWithSmallAudience() {
        Map<String, Play> plays = Map.of("hamlet", Play.createPlay("Hamlet", Play.PlayType.tragedy));
        Customer customer = new Customer("BigCo", 12345, 100); // Create a Customer object
        Invoice invoice = new Invoice(customer, List.of(new Performance("hamlet", 10)));
        StatementPrinter statementPrinter = new StatementPrinter();

        var result = statementPrinter.print(invoice, plays);
        
        var result2 = InvoiceToHTML.generateHTMLInvoice(invoice.getCustomer().getName(), invoice.performances, plays); // Use the updated method with Invoice object
        String expected = "Statement for BigCo\n" +
                          "  Hamlet: $400.00 (10 seats)\n" +
                          "Amount owed is $400.00\n" +
                          "You earned 0 credits\n";

        String expectedHTML = "<html> <style> table, th, td {border:1px solid black;} </style> <head><title>Facture</title></head><body><h1>Invoice</h1><ul><li><b>Client : </b>BigCo</li></ul><table><tr><th>Piece</th><th>Seats sold</th><th>Price</th></tr><tr><td>Hamlet</td><td>10</td><td>$400.00</td></tr><tr><th colspan=\"2\">Total owed : </th><td>$400.00</td></tr><tr><th colspan=\"2\">Fidelity points earned : </th><td>0</td></tr></table></body></html>";

        assertEquals(expected, result);
        assertEquals(expectedHTML, result2);
    }

    @Test
    void comedyWithSmallAudience() {
        Map<String, Play> plays = Map.of("As You Like It", Play.createPlay("As You Like It", Play.PlayType.comedy));
        Customer customer = new Customer("BigCo", 12345, 100); // Create a Customer object
        Invoice invoice = new Invoice(customer, List.of(new Performance("As You Like It", 10)));
        StatementPrinter statementPrinter = new StatementPrinter();

        var result = statementPrinter.print(invoice, plays);
        var result2 = InvoiceToHTML.generateHTMLInvoice(invoice.getCustomer().getName(), invoice.performances, plays);

        String expected = "Statement for BigCo\n" +
                          "  As You Like It: $330.00 (10 seats)\n" +
                          "Amount owed is $330.00\n" +
                          "You earned 2 credits\n";
        
        String expectedHTML = "<html> <style> table, th, td {border:1px solid black;} </style> <head><title>Facture</title></head><body><h1>Invoice</h1><ul><li><b>Client : </b>BigCo</li></ul><table><tr><th>Piece</th><th>Seats sold</th><th>Price</th></tr><tr><td>As You Like It</td><td>10</td><td>$330.00</td></tr><tr><th colspan=\"2\">Total owed : </th><td>$330.00</td></tr><tr><th colspan=\"2\">Fidelity points earned : </th><td>2</td></tr></table></body></html>";

        assertEquals(expected, result);
        assertEquals(expectedHTML, result2);
    }




    @Test
    void exampleStatement2() {
        Map<String, Play> plays = Map.of(
                "hamlet",  Play.createPlay("Hamlet", Play.PlayType.tragedy),
                "as-like", Play.createPlay("As You Like It", Play.PlayType.comedy),
                "othello", Play.createPlay("Othello", Play.PlayType.tragedy));

        Customer customer = new Customer("BigCo", 12345, 130); // Create a Customer object

        Invoice invoice = new Invoice(customer, List.of(
                new Performance("hamlet", 55),
                new Performance("as-like", 35),
                new Performance("othello", 40)));

        StatementPrinter statementPrinter = new StatementPrinter();
        var result = statementPrinter.print(invoice, plays);
        
        String expected = "Statement for BigCo\n" +
                        "  Hamlet: $650.00 (55 seats)\n" +
                        "  As You Like It: $580.00 (35 seats)\n" +
                        "  Othello: $500.00 (40 seats)\n" +
                        "Amount owed is $1,730.00\n" +
                        "You earned 47 credits\n" +
                        "You have 177 Loyalty Points, you get a $15 reduction\n" +
                        "Amount owed after reduction $1,715.00\n" +
                        "New Fidelity Points 27\n";
        assertEquals(expected, result);
    }


    



    


    @Test
    void statementWithNewPlayTypes() {
        Map<String, Play> plays = Map.of(
                "henry-v",  Play.createPlay("Henry V", Play.PlayType.history),
                "as-like", Play.createPlay("As You Like It", Play.PlayType.pastoral));
    
        Customer customer = new Customer("BigCo", 12345, 100); 
    
        Invoice invoice = new Invoice(customer, List.of(
                new Performance("henry-v", 53),
                new Performance("as-like", 55)));
    
        StatementPrinter statementPrinter = new StatementPrinter();
        Assertions.assertThrows(Error.class, () -> {
            statementPrinter.print(invoice, plays);
        });
        Assertions.assertThrows(Error.class, () -> {
            InvoiceToHTML.generateHTMLInvoice(invoice.getCustomer().getName(), invoice.performances, plays);
        });
    }
    
}
