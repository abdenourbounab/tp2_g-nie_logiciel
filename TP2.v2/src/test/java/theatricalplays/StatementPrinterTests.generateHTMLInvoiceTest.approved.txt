<html> <style> table, th, td {border:1px solid black;} </style> <head><title>Facture</title></head><body><h1>Invoice</h1><ul><li><b>Client : </b>BigCo</li></ul><table><tr><th>Piece</th><th>Seats sold</th><th>Price</th></tr><tr><td>Hamlet</td><td>55</td><td>$650.00</td></tr><tr><td>As You Like It</td><td>35</td><td>$580.00</td></tr><tr><td>Othello</td><td>40</td><td>$500.00</td></tr><tr><th colspan="2">Total owed : </th><td>$1,730.00</td></tr><tr><th colspan="2">Fidelity points earned : </th><td>47</td></tr></table></body></html>